<a href="https://colab.research.google.com/github/mateosuster/pythonungs/blob/master/codigos/TPs/MPE3_022023_TP4_resolucion.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Matemática para Economistas III - UNGS

## TP N°4 - Posible resolución


#### 1) Cargue y guarde en una variable el dataset de Reservas Internacionales (promedio mensual) disponibilizado por el BCRA

In [1]:
#importamos las librerías, siempre al principio
import pandas as pd
import numpy as np

In [2]:
#cargamos el dataset
df = pd.read_csv("https://infra.datos.gob.ar/catalog/sspm/dataset/92/distribution/92.2/download/reservas-internacionales-pasivos-financieros-bcra.csv")

#### 2) Explore el dataset con las funciones correspondientes. Detalle a grandes rasgos qué información poseen sus columnas y de qué tipos de datos son las mismas.

El dataset posee información sobre las reservas internacionales del BCRA, de algunos de sus pasivos e incluso sobre el tipo de cambio con el que se valuan los mismos. El método `.info()` nos brinda un resumen sintético de la información del dataset: columnas, cantidad de valores no nulos, tipo de dato de cada una de ellas, etc.



In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7608 entries, 0 to 7607
Data columns (total 12 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   indice_tiempo                      7608 non-null   object 
 1   reservas_internacionales_dolares   7608 non-null   float64
 2   pasivos_monetarios                 7608 non-null   float64
 3   pasivos_monetarios_letras_pesos    7608 non-null   float64
 4   pasivos_monetarios_letras_dolares  7608 non-null   float64
 5   letras_liquidez_saldo_pesos        7608 non-null   float64
 6   posicion_neta_pases                7608 non-null   float64
 7   otros_depositos_gobierno           7608 non-null   float64
 8   redescuentos_adelantos_iliquidez   7608 non-null   float64
 9   tipo_cambio_valuacion              7608 non-null   float64
 10  pos_neta_det_des_pases_pasivos     7608 non-null   float64
 11  pos_neta_det_des_pases_activos     7608 non-null   float

 La columna que contiene la fecha, `indice_tiempo`, es de tipo object, lo cual indica que es un string, mientras que el resto son de tipo float, es decir, números reales. Una alternativa es el atributo `.dtypes`

In [7]:
df.dtypes

indice_tiempo                         object
reservas_internacionales_dolares     float64
pasivos_monetarios                   float64
pasivos_monetarios_letras_pesos      float64
pasivos_monetarios_letras_dolares    float64
letras_liquidez_saldo_pesos          float64
posicion_neta_pases                  float64
otros_depositos_gobierno             float64
redescuentos_adelantos_iliquidez     float64
tipo_cambio_valuacion                float64
pos_neta_det_des_pases_pasivos       float64
pos_neta_det_des_pases_activos       float64
dtype: object


El método `.head()` nos permite ver el encabezado del dataset, lo cual es más prolijo que imprimir el dataset entero por la pantalla.

In [4]:
df.head()

indice_tiempo  reservas_internacionales_dolares  pasivos_monetarios  \
0    2003-01-02                      10500.743384             30300.0   
1    2003-01-03                      10505.278810             30359.0   
2    2003-01-04                      10505.278810             30359.0   
3    2003-01-05                      10505.278810             30359.0   
4    2003-01-06                      10563.790206             30538.0   

   pasivos_monetarios_letras_pesos  pasivos_monetarios_letras_dolares  \
0                      2584.292106                         558.358408   
1                      2650.113332                         643.514769   
2                      2650.113332                         643.514769   
3                      2650.113332                         643.514769   
4                      2650.113332                         637.830787   

   letras_liquidez_saldo_pesos  posicion_neta_pases  otros_depositos_gobierno  \
0                          0.0               -187.0                      67.0   
1                          0.0               -187.0                      48.0   
2                          0.0               -187.0                      48.0   
3                          0.0               -187.0                      48.0   
4                          0.0               -187.0                      46.0   

   redescuentos_adelantos_iliquidez  tipo_cambio_valuacion  \
0                           20176.0                 3.3630   
1                           20176.0                 3.3625   
2                           20176.0                 3.3625   
3                           20176.0                 3.3625   
4                           20176.0                 3.3328   

   pos_neta_det_des_pases_pasivos  pos_neta_det_des_pases_activos  
0                             0.0                           187.0  
1                             0.0                           187.0  
2                             0.0                           187.0  
3                             0.0                           187.0  
4                             0.0                           187.0

El atributo `.shape` nos permite ver la forma del dataset, el cual posee 7608 filas y 12 columnas.

In [5]:
df.shape

(7608, 12)

A su vez, `.size` nos da el tamaño de la matriz

In [6]:
df.size

91296

#### 3) Convierta la variable "indice_tiempo" en una variable tipo fecha. ¿Cuál es el rango de la variable?

In [11]:
#conversion
df["indice_tiempo"] = pd.to_datetime(df["indice_tiempo"])

#busco valores minimos y maximos de la variable, para ver su rango
print(f'El valor mínimo es {df["indice_tiempo"].min()}')
print(f'El valor máximo es {df["indice_tiempo"].max()}')

El valor mínimo es 2003-01-02 00:00:00
El valor máximo es 2023-10-31 00:00:00


#### 4) Filtre el dataset para quedarse únicamente con los datos entre enero de 2010 a diciembre de 2020. Utilice el método .loc

In [14]:
#se fijan los años que se quieren filtrar en variables
start_year = 2010
end_year = 2020

#se construye una máscara booleana: prestar atención al atributo .dt.year que devuelve el año de la fecha, facilitando la manipulacion
boolean_mask = (df["indice_tiempo"].dt.year >= start_year) & (df["indice_tiempo"].dt.year <= end_year)

#se pasa la máscara en el dataframe
df_filtrado = df.loc[boolean_mask]

#se corrobora su tamaño
df_filtrado.shape

(4018, 12)

#### 5) Cree una nueva columna que contenga la variación absoluta de la variable de reservas internacionales entre períodos. Previo a eso, para asegurarse que el cálculo está bien hecho, ordene el dataset por fecha.

In [15]:
#ordeno y asigno el dataset ordenado
df_filtrado = df_filtrado.sort_index()

In [18]:
# alternativa 1: con método .diff()
df_filtrado['variacion_abs_rrii_1'] = df_filtrado["reservas_internacionales_dolares"].diff()

#muestro sus valores estadisticos
df_filtrado['variacion_abs_rrii_1'].describe()

count     4017.000000
mean        -2.136041
std        502.723829
min      -4749.000000
25%        -38.000000
50%          0.000000
75%         10.000000
max      14796.200000
Name: variacion_abs_rrii, dtype: float64

In [38]:
# alternativa 2: con función propia

def variacion_absoluta(x):
  return x - x.shift(1)

df_filtrado['variacion_abs_rrii_2'] = variacion_absoluta(df_filtrado["reservas_internacionales_dolares"])
df_filtrado['variacion_abs_rrii_2'].describe()

count     4017.000000
mean        -2.136041
std        502.723829
min      -4749.000000
25%        -38.000000
50%          0.000000
75%         10.000000
max      14796.200000
Name: variacion_abs_rrii_2, dtype: float64